src/dragonball/src/resource_manager.rs

```diff
@@ -0,0 +1,683 @@
+                    match self.allocate_legacy_irq(shared_irq, *irq) {
+                        Some(v) => Resource::LegacyIrq(v),
+                        None => return self.free_allocated_resources(resources),
+                    }
+                }
+                ResourceConstraint::PciMsiIrq { size } => {
+                    match self.allocate_msi_irq_aligned(*size) {
+                        Some(base) => Resource::MsiIrq {
+                            ty: MsiIrqType::PciMsi,
+                            base,
+                            size: *size,
+                        },
+                        None => return self.free_allocated_resources(resources),
+                    }
+                }
+                ResourceConstraint::PciMsixIrq { size } => match self.allocate_msi_irq(*size) {
+                    Some(base) => Resource::MsiIrq {
+                        ty: MsiIrqType::PciMsix,
+                        base,
+                        size: *size,
+                    },
+                    None => return self.free_allocated_resources(resources),
+                },
+                ResourceConstraint::GenericIrq { size } => match self.allocate_msi_irq(*size) {
+                    Some(base) => Resource::MsiIrq {
+                        ty: MsiIrqType::GenericMsi,
+                        base,
+                        size: *size,
+                    },
+                    None => return self.free_allocated_resources(resources),
+                },
+                ResourceConstraint::KvmMemSlot { slot, size } => {
+                    match self.allocate_kvm_mem_slot(*size, *slot) {
+                        Some(v) => Resource::KvmMemSlot(v),
+                        None => return self.free_allocated_resources(resources),
+                    }
+                }
+            };
+            resources.append(res);
+        }
+
+        Ok(resources)
+    }
+
+    /// Free resources allocated for a device.
+    pub fn free_device_resources(&self, resources: &DeviceResources) {
+        for res in resources.iter() {
+            match res {
+                Resource::PioAddressRange { base, size } => self.free_pio_address(*base, *size),
+                Resource::MmioAddressRange { base, size } => self.free_mmio_address(*base, *size),
+                Resource::MemAddressRange { base, size } => self.free_mem_address(*base, *size),
+                Resource::LegacyIrq(base) => self.free_legacy_irq(*base),
+                Resource::MsiIrq { ty: _, base, size } => self.free_msi_irq(*base, *size),
+                Resource::KvmMemSlot(slot) => self.free_kvm_mem_slot(*slot),
+                Resource::MacAddresss(_) => {}
+            }
+        }
+    }
+
+    fn free_allocated_resources(
+        &self,
+        resources: DeviceResources,
+    ) -> Result<DeviceResources, ResourceError> {
+        self.free_device_resources(&resources);
+        Err(ResourceError::NoAvailResource)
```

Comment:
This is slightly odd code: All the `free_*()` handlers `panic` on error and the function that calls the handlers then returns an error "on success". Also, if the resources are freed, saying there are no resources available is "correct", but it's potentially misleading as it really means, "there (probably) _are_ system resources to _create_ the requested resources, but those requested resources haven't been (re-)created yet".

Why can we just use the conventional methodology here and make the handlers return a `Result` to allow `free_device_resources()` to return that `Result`. This function could then return that `Result` *on error*, or `Ok` on success, as is normal.

Reply:
Thanks for your suggestion.

We have refactored this part.